File: backend/app/services/locales.py
"""Locale handling shared by UI and the editor.

`SUPPORTED` is the catalog of selectable UI locales (code, native name, direction).
`co_lang` maps an app locale to the language code the editor expects in the `lang`
URL parameter.
"""

# code -> (native name, direction)
SUPPORTED: dict[str, tuple[str, str]] = {
    "en": ("English", "ltr"),
    "es": ("Español", "ltr"),
    "de": ("Deutsch", "ltr"),
    "fr": ("Français", "ltr"),
    "pt-BR": ("Português (Brasil)", "ltr"),
    "ru": ("Русский", "ltr"),
    "it": ("Italiano", "ltr"),
    "nl": ("Nederlands", "ltr"),
    "pl": ("Polski", "ltr"),
    "uk": ("Українська", "ltr"),
    "tr": ("Türkçe", "ltr"),
    "cs": ("Čeština", "ltr"),
    "zh-CN": ("简体中文", "ltr"),
    "ja": ("日本語", "ltr"),
    "ko": ("한국어", "ltr"),
    "hi": ("हिन्दी", "ltr"),
    "vi": ("Tiếng Việt", "ltr"),
    "id": ("Bahasa Indonesia", "ltr"),
    "th": ("ไทย", "ltr"),
    "ar": ("العربية", "rtl"),
    "he": ("עברית", "rtl"),
    "fa": ("فارسی", "rtl"),
}

# App locale -> editor `lang` code (normalize a few that differ).
_CO_OVERRIDES: dict[str, str] = {
    "en": "en-US",
    "ru": "ru-RU",
    "uk": "uk-UA",
    "cs": "cs-CZ",
    "ja": "ja-JP",
    "ko": "ko-KR",
    "hi": "hi-IN",
    "th": "th-TH",
    "he": "he-IL",
    "fa": "fa-IR",
    "vi": "vi-VN",
    "id": "id-ID",
}


def is_supported(code: str) -> bool:
    return code in SUPPORTED


def direction(code: str) -> str:
    return SUPPORTED.get(code, ("", "ltr"))[1]


def co_lang(code: str) -> str:
    """Return the editor-acceptable language tag for an app locale."""
    if code in _CO_OVERRIDES:
        return _CO_OVERRIDES[code]
    return code

File: frontend/src/i18n/locales.ts
// Selectable UI locales. Mirrors backend app/services/locales.py SUPPORTED.
export interface LocaleInfo {
  code: string;
  native: string;
  dir: "ltr" | "rtl";
}

export const LOCALES: LocaleInfo[] = [
  { code: "en", native: "English", dir: "ltr" },
  { code: "es", native: "Español", dir: "ltr" },
  { code: "de", native: "Deutsch", dir: "ltr" },
  { code: "fr", native: "Français", dir: "ltr" },
  { code: "pt-BR", native: "Português (Brasil)", dir: "ltr" },
  { code: "ru", native: "Русский", dir: "ltr" },
  { code: "it", native: "Italiano", dir: "ltr" },
  { code: "nl", native: "Nederlands", dir: "ltr" },
  { code: "pl", native: "Polski", dir: "ltr" },
  { code: "uk", native: "Українська", dir: "ltr" },
  { code: "tr", native: "Türkçe", dir: "ltr" },
  { code: "cs", native: "Čeština", dir: "ltr" },
  { code: "zh-CN", native: "简体中文", dir: "ltr" },
  { code: "ja", native: "日本語", dir: "ltr" },
  { code: "ko", native: "한국어", dir: "ltr" },
  { code: "hi", native: "हिन्दी", dir: "ltr" },
  { code: "vi", native: "Tiếng Việt", dir: "ltr" },
  { code: "id", native: "Bahasa Indonesia", dir: "ltr" },
  { code: "th", native: "ไทย", dir: "ltr" },
  { code: "ar", native: "العربية", dir: "rtl" },
  { code: "he", native: "עברית", dir: "rtl" },
  { code: "fa", native: "فارسی", dir: "rtl" },
];

export const LOCALE_CODES = LOCALES.map((l) => l.code);

export function dirFor(code: string): "ltr" | "rtl" {
  return LOCALES.find((l) => l.code === code)?.dir ?? "ltr";
}

export function nativeName(code: string): string {
  return LOCALES.find((l) => l.code === code)?.native ?? code;
}
